src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsMPM.cpp

```diff
@@ -0,0 +1,897 @@
+                    "However if a new surface is generated, then the mechanics solve must be executed again due to the "
+                    "change in topology." );
+
+  registerWrapper( viewKeyStruct::strainTheoryString(), &m_strainTheory ).
+    setApplyDefaultValue( 0 ).
+    setInputFlag( InputFlags::OPTIONAL ).
+    setDescription( "Indicates whether or not to use "
+                    "`Infinitesimal Strain Theory <https://en.wikipedia.org/wiki/Infinitesimal_strain_theory>`_, or "
+                    "`Finite Strain Theory <https://en.wikipedia.org/wiki/Finite_strain_theory>`_. Valid Inputs are:\n"
+                    " 0 - Infinitesimal Strain \n"
+                    " 1 - Finite Strain" );
+
+  registerWrapper( viewKeyStruct::contactRelationNameString(), &m_contactRelationName ).
+    setApplyDefaultValue( viewKeyStruct::noContactRelationNameString() ).
+    setInputFlag( InputFlags::OPTIONAL ).
+    setDescription( "Name of contact relation to enforce constraints on fracture boundary." );
+
+  registerWrapper( viewKeyStruct::maxForceString(), &m_maxForce ).
+    setInputFlag( InputFlags::FALSE ).
+    setDescription( "The maximum force contribution in the problem domain." );
+
+}
+
+void SolidMechanicsMPM::postProcessInput()
+{
+  SolverBase::postProcessInput();
+
+//  checkModelNames( m_solidMaterialNames, viewKeyStruct::solidMaterialNamesString() );
+
+  LinearSolverParameters & linParams = m_linearSolverParameters.get();
+  linParams.isSymmetric = true;
+  linParams.dofsPerNode = 3;
+  linParams.amg.separateComponents = true;
+}
+
+SolidMechanicsMPM::~SolidMechanicsMPM()
+{
+  // TODO Auto-generated destructor stub
+}
+
+
+void SolidMechanicsMPM::registerDataOnMesh( Group & meshBodies )
+{
+  ExecutableGroup::registerDataOnMesh( meshBodies );
+
+  forMeshTargets( meshBodies, [&] ( string const &,
+                                    MeshLevel & meshLevel,
+                                    arrayView1d< string const > const & regionNames )
+  {
+    ParticleManager & particleManager = meshLevel.getParticleManager();
+
+    MeshBody const & meshBody = dynamicCast< MeshBody const & >( meshLevel.getParent().getParent() );
+
+    // Set constitutive names on particles
+    if(meshBody.m_hasParticles)
+    {
+      particleManager.forParticleSubRegions< ParticleSubRegionBase >( regionNames,
+                                                                      [&]( localIndex const,
+                                                                          ParticleSubRegionBase & subRegion )
+      {
+        setConstitutiveNamesCallSuper( subRegion );
+        setConstitutiveNames( subRegion );
+      } );
+    }
+
+  } );
+
+  forMeshTargets( meshBodies, [&] ( string const &,
+                                    MeshLevel & meshLevel,
+                                    arrayView1d<string const> const & regionNames )
+  {
+    MeshBody const & meshBody = dynamicCast< MeshBody const & >( meshLevel.getParent().getParent() );
```

Comment:
Could you iterate on the `meshBodies` in a classical way and then access the `meshLevels`?
Tricking with `getParent().getParent()` sounds like this is not the pattern awaited by the fwk.

Reply:
Now using `MeshBody const & meshBody = meshBodies.getGroup< MeshBody >( meshBodyName );` since `meshBodyName` is an input to this lambda (previously unused). This way I can keep the `forMeshTargets` loop.